hawkular-wildfly-agent/src/main/java/org/hawkular/agent/monitor/protocol/EndpointService.java

```diff
@@ -237,9 +260,14 @@ public void addInventoryListener(InventoryListener listener) {
      * @param listener to remove
      */
     public void removeInventoryListener(InventoryListener listener) {
-        status.assertInitialOrStopped(getClass(), "removeInventoryListener()");
-        this.inventoryListenerSupport.inventoryListeners.remove(listener);
-        LOG.debugf("Removed inventory listener [%s] for endpoint [%s]", listener, getMonitoredEndpoint());
+        try {
+            this.inventoryListenerSupport.inventoryListenerRWLock.writeLock().lock();
```

Comment:
acquisition of all locks should always come outside (just before) entering the try-catch. Because if something goes wrong, the finally block is going to try to unlock a lock you have not acquired. So swap lines 263-264. Do the same for the others.